### ISSUE TITLE
Value_select to perform region-wise selection

### ISSUE THREAD

**Issue by Naruto-Sasuke** [pytorch/pytorch#24313]
## 🚀 Feature
```python
import torch

a = torch.tensor([[0,0,1,1,1],[0,0,1,1,2],[3,3,3,3,2]])
c = 2
feat = torch.randn(c, *a.size())
num_sp = a.max().long() + 1

# For now, I have to use torch.masked_select for many times
output = []
for i in range(num_sp):
    # (c*1), mean of each SP.
    output.append(torch.masked_select(feat, a==i).view(c, -1).mean(-1, keepdim=True)) 
output = torch.stack(output)
print('feat')
print(feat)
print('ref_mask')
print(ref_mask)
print('output')
print(output)

# Something like this will be better
output, return_index = torch.value_select(feat, mask_ref = a, values=torch.arange(num_sp))

```
```bash
feat
tensor([[[-0.6140, -1.5337,  1.4974, -1.2668, -0.3866],
         [-0.1943,  0.0868,  1.0380,  0.1135,  1.1820],
         [-1.4744,  1.5889, -0.2627, -1.7951, -0.0045]],

        [[-0.8168, -0.6949, -1.9554, -0.5807, -0.2200],
         [-0.7170, -0.8191,  0.5452,  1.6736, -0.6598],
         [-0.6220,  0.8880, -0.6597,  0.6438, -1.5148]]])
ref_mask
tensor([[0, 0, 1, 1, 1],
        [0, 0, 1, 1, 2],
        [3, 3, 3, 3, 2]])
output
tensor([[[-0.5638],
         [-0.7619]],

        [[ 0.1991],
         [-0.1075]],

        [[ 0.5887],
         [-1.0873]],

        [[-0.4858],
         [ 0.0625]]])
```

## Motivation
In some cases, label matrix (ref_mask above) indicates which class the pixel should be categorized to. Suppose we get some labels from a super-pixel algorithm, then I want to calculate the mean of SP region or perform other operations for each region. Then `value_select` function is quite useful. In a word, it performs `region-wise` selection and will be useful for other tasks too.

## Pitch
As the pixels contained in each region is not constant, then we need to return the indexes for each region, logged in `return_index`.
